contracts/distribution/dao-rewards-distributor/src/tests.rs

```diff
@@ -0,0 +1,2175 @@
+
+    app.borrow_mut().update_block(next_block);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR1, 2000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR2, 1000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR3, 1000);
+
+    assert_eq!(get_balance_native(&app, ADDR1, &denom), Uint128::zero());
+    claim_rewards(&mut app, reward_addr.clone(), ADDR1);
+    assert_eq!(get_balance_native(&app, ADDR1, &denom), Uint128::new(2000));
+    assert_pending_rewards(&mut app, &reward_addr, ADDR1, 0);
+
+    app.borrow_mut().update_block(|b| b.height += 10);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR1, 5000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR2, 3500);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR3, 3500);
+
+    unstake_nft(&mut app, &vp_addr, ADDR2, "3");
+    unstake_nft(&mut app, &vp_addr, ADDR3, "4");
+
+    app.borrow_mut().update_block(|b| b.height += 10);
+
+    assert_pending_rewards(&mut app, &reward_addr, ADDR1, 15000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR2, 3500);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR3, 3500);
+
+    claim_rewards(&mut app, reward_addr.clone(), ADDR1);
+    assert_eq!(get_balance_native(&app, ADDR1, &denom), Uint128::new(17000));
+
+    claim_rewards(&mut app, reward_addr.clone(), ADDR2);
+    assert_eq!(get_balance_native(&app, ADDR2, &denom), Uint128::new(3500));
+
+    stake_nft(&mut app, &vp_addr, &cw721_addr, ADDR2, "3");
+    stake_nft(&mut app, &vp_addr, &cw721_addr, ADDR3, "4");
+
+    app.borrow_mut().update_block(|b| b.height += 10);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR1, 5000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR2, 2500);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR3, 6000);
+
+    // Current height is 1034. ADDR1 is receiving 500 tokens/block
+    // and ADDR2 / ADDR3 are receiving 250.
+    //
+    // At height 101000 99966 additional blocks have passed. So we
+    // expect:
+    //
+    // ADDR1: 5000 + 99966 * 500 = 49,998,000
+    // ADDR2: 2500 + 99966 * 250 = 24,994,000
+    // ADDR3: 6000 + 99966 * 250 = 24,997,500
+    app.borrow_mut().update_block(|b| b.height = 101000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR1, 49988000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR2, 24994000);
+    assert_pending_rewards(&mut app, &reward_addr, ADDR3, 24997500);
+
+    claim_rewards(&mut app, reward_addr.clone(), ADDR1);
+    claim_rewards(&mut app, reward_addr.clone(), ADDR2);
+    assert_eq!(
+        get_balance_native(&app, ADDR1, &denom),
+        Uint128::new(50005000)
+    );
+    assert_eq!(
+        get_balance_native(&app, ADDR2, &denom),
+        Uint128::new(24997500)
+    );
+    assert_eq!(get_balance_native(&app, ADDR3, &denom), Uint128::new(0));
+    assert_eq!(
+        get_balance_native(&app, &reward_addr, &denom),
+        Uint128::new(24997500)
+    );
+}
+
+#[test]
+fn test_cw4_dao_rewards() {
```

Comment:
This test is sadly not working currently, and I'm not quite sure why. :smiling_face_with_tear:

Reply:
Fixed! Noticed some interesting behavior around snapshot maps, messages, and submessages.

The fix was to specify the currently block height when querying `VotingPowerAtHeight` and `TotalPowerAtHeight`.